File: apps/desktop/lib/widgets/agent_actions_sheet.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:provider/provider.dart';
import '../design/nano_tokens.dart';
import '../models/models.dart';
import '../providers/app_provider.dart';
import '../services/server_service.dart';
import 'nano/nano_card.dart';

/// Bottom sheet of quick actions for an agent: open terminal, request fresh
/// data (POST /data-request), copy id, and reboot host (POST /command →
/// SYSTEM_REBOOT). The data/command actions hit the backend directly via the
/// agent's [ServerService] and surface the real execution outcome through a
/// progress → result snackbar (with a dismiss action), instead of fire-and-
/// forget.
Future<void> showAgentActionsSheet(
  BuildContext context, {
  required Agent agent,
  VoidCallback? onOpenTerminal,
}) {
  final t = context.nano;
  return showModalBottomSheet(
    context: context,
    backgroundColor: t.card,
    showDragHandle: true,
    shape: RoundedRectangleBorder(
      borderRadius:
          BorderRadius.vertical(top: Radius.circular(t.isIOS ? 16 : 28)),
    ),
    builder: (ctx) {
      return SafeArea(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Padding(
              padding: const EdgeInsets.fromLTRB(20, 0, 20, 8),
              child: Align(
                alignment: Alignment.centerLeft,
                child: Text('actions.title'.tr(),
                    style: TextStyle(
                        fontSize: 17,
                        fontWeight: FontWeight.w600,
                        color: t.fg)),
              ),
            ),
            if (agent.permissionLevel > 0)
              _ActionRow(
                icon: Icons.terminal_rounded,
                label: 'actions.openTerminal'.tr(),
                sub: 'actions.openTerminalSub'.tr(),
                onTap: () {
                  Navigator.pop(ctx);
                  onOpenTerminal?.call();
                },
              ),
            _ActionRow(
              icon: Icons.refresh_rounded,
              label: 'actions.requestData'.tr(),
              sub: 'actions.requestDataSub'.tr(),
              onTap: () {
                final messenger = ScaffoldMessenger.of(context);
                final svc = context.read<AppProvider>().serviceForAgent(agent.id);
                Navigator.pop(ctx);
                if (svc == null) {
                  _show(messenger, 'actions.noServerForNode'.tr());
                  return;
                }
                _runRequestData(messenger, svc, agent.id);
              },
            ),
            _ActionRow(
              icon: Icons.copy_rounded,
              label: 'actions.copyAgentId'.tr(),
              onTap: () {
                Clipboard.setData(ClipboardData(text: agent.id));
                final messenger = ScaffoldMessenger.of(context);
                Navigator.pop(ctx);
                _show(messenger, 'actions.agentIdCopied'.tr());
              },
            ),
            if (agent.permissionLevel >= 3)
              _ActionRow(
                icon: Icons.power_settings_new_rounded,
                label: 'actions.reboot'.tr(),
                sub: 'actions.rebootSub'.tr(),
                danger: true,
                onTap: () async {
                  final messenger = ScaffoldMessenger.of(context);
                  final provider = context.read<AppProvider>();
                  Navigator.pop(ctx);
                  final ok = await _confirmReboot(context);
                  if (ok != true) return;
                  final svc = provider.serviceForAgent(agent.id);
                  if (svc == null) {
                    _show(messenger, 'actions.noServerForNode'.tr());
                    return;
                  }
                  _runReboot(messenger, svc, agent.id);
                },
              ),
            const SizedBox(height: 8),
          ],
        ),
      );
    },
  );
}

/// Plain transient snackbar (copy id / no-server errors).
void _show(ScaffoldMessengerState messenger, String msg) {
  messenger.showSnackBar(SnackBar(content: Text(msg)));
}

/// Persistent progress snackbar with a spinner — shown while a command is
/// dispatched and polled. Returns nothing; replaced by [_showResult] when done.
void _showProgress(ScaffoldMessengerState messenger, String msg) {
  messenger
    ..hideCurrentSnackBar()
    ..showSnackBar(SnackBar(
      duration: const Duration(minutes: 1),
      content: Row(
        children: [
          const SizedBox(
            width: 16,
            height: 16,
            child: CircularProgressIndicator(strokeWidth: 2),
          ),
          const SizedBox(width: 14),
          Expanded(child: Text(msg)),
        ],
      ),
    ));
}

/// Final result snackbar — carries a dismiss action ('撤销'/close), per the
/// MDSnack design (android-app.jsx L790-794).
void _showResult(ScaffoldMessengerState messenger, String msg) {
  messenger
    ..hideCurrentSnackBar()
    ..showSnackBar(SnackBar(
      duration: const Duration(seconds: 6),
      content: Text(msg),
      action: SnackBarAction(
        label: 'common.dismiss'.tr(),
        onPressed: messenger.hideCurrentSnackBar,
      ),
    ));
}

/// Request fresh data via the data-request endpoint, surfacing the real outcome
/// through a progress → result snackbar.
Future<void> _runRequestData(
  ScaffoldMessengerState messenger,
  ServerService svc,
  String agentId,
) async {
  _showProgress(messenger, 'actions.requestDataPending'.tr());
  final err = await svc.requestData(agentId);
  _showResult(
    messenger,
    err == null
        ? 'actions.dataRequested'.tr()
        : 'actions.dataRequestFailed'.tr(namedArgs: {'error': err}),
  );
}

/// Dispatch SYSTEM_REBOOT and poll its structured result so the snackbar
/// reflects the agent's actual execution (pending → ready / denied / error)
/// rather than just the dispatch ack.
Future<void> _runReboot(
  ScaffoldMessengerState messenger,
  ServerService svc,
  String agentId,
) async {
  _showProgress(messenger, 'actions.rebootPending'.tr());
  final dispatch = await svc.sendCommandReturningId(agentId, 'SYSTEM_REBOOT');
  if (!dispatch.ok) {
    _showResult(
      messenger,
      'actions.rebootFailed'.tr(namedArgs: {'error': dispatch.error ?? ''}),
    );
    return;
  }
  final result = await _pollUntilDone(svc, agentId, dispatch.commandId!);
  switch (result.status) {
    case CommandResultStatus.ready:
      _showResult(messenger, 'actions.rebootDone'.tr());
      break;
    case CommandResultStatus.denied:
      _showResult(messenger, 'actions.commandDenied'.tr());
      break;
    case CommandResultStatus.pending:
      // Reboot disconnects the host before it can report — treat as sent.
      _showResult(messenger, 'actions.rebootSent'.tr());
      break;
    case CommandResultStatus.error:
      _showResult(
        messenger,
        'actions.rebootFailed'.tr(namedArgs: {'error': result.message ?? ''}),
      );
      break;
  }
}

/// Poll a dispatched command's result with a short backoff until it leaves the
/// pending state or the attempt budget is exhausted (returns the last pending
/// result so callers can decide how to render it).
Future<CommandResult> _pollUntilDone(
  ServerService svc,
  String agentId,
  String commandId, {
  int maxAttempts = 6,
  Duration interval = const Duration(seconds: 1),
}) async {
  CommandResult result = const CommandResult(CommandResultStatus.pending);
  for (var i = 0; i < maxAttempts; i++) {
    result = await svc.pollCommandResult(agentId, commandId);
    if (!result.isPending) return result;
    await Future<void>.delayed(interval);
  }
  return result;
}

Future<bool?> _confirmReboot(BuildContext context) {
  final t = context.nano;
  return showDialog<bool>(
    context: context,
    builder: (ctx) => AlertDialog(
      backgroundColor: t.card,
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(t.isIOS ? 14 : 28),
      ),
      title: Text('actions.rebootConfirmTitle'.tr(),
          style: TextStyle(color: t.fg)),
      content: Text(
        'actions.rebootConfirmBody'.tr(),
        style: TextStyle(color: t.fg2),
      ),
      actions: [
        TextButton(
          onPressed: () => Navigator.pop(ctx, false),
          child: Text('common.cancel'.tr(), style: TextStyle(color: t.fg2)),
        ),
        // Filled danger button (bg crit, white text) per MDDialog
        // filled-danger primary action (android-app.jsx L765-787).
        FilledButton(
          onPressed: () => Navigator.pop(ctx, true),
          style: FilledButton.styleFrom(
            backgroundColor: t.crit,
            foregroundColor: Colors.white,
            shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(t.isIOS ? 10 : 20),
            ),
          ),
          child: Text('actions.rebootConfirm'.tr()),
        ),
      ],
    ),
  );
}

class _ActionRow extends StatelessWidget {
  final IconData icon;
  final String label;
  final String? sub;
  final bool danger;
  final VoidCallback onTap;
  const _ActionRow({
    required this.icon,
    required this.label,
    this.sub,
    this.danger = false,
    required this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    final t = context.nano;
    final color = danger ? t.crit : t.fg;
    return NanoListRow(
      divider: false,
      onTap: onTap,
      padding: const EdgeInsets.fromLTRB(20, 14, 20, 14),
      leading: Icon(icon, color: danger ? t.crit : t.accent, size: 22),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(label,
              style: TextStyle(
                  fontSize: 15, fontWeight: FontWeight.w500, color: color)),
          if (sub != null)
            Text(sub!, style: TextStyle(fontSize: 12, color: t.fg3)),
        ],
      ),
    );
  }
}

File: apps/ios/NanoLink/Sheets/AgentActionsSheet.swift
import SwiftUI
import UIKit

/// Permission-aware actions for a node: terminal, metrics refresh, ID copy and
/// L3-only reboot with command-result polling.
struct AgentActionsSheet: View {
    let agent: Agent
    var onOpenTerminal: (() -> Void)? = nil

    @EnvironmentObject private var store: AppStore
    @Environment(\.dismiss) private var dismiss
    @Environment(\.nano) private var t
    @State private var confirmReboot = false
    @State private var busyMessage: String?
    @State private var resultMessage: String?
    @State private var resultIsError = false

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                NanoCard {
                    if agent.permissionLevel >= 3 {
                        actionRow("terminal", "actions.openTerminal", "actions.openTerminalSub") {
                            dismiss(); onOpenTerminal?()
                        }
                    }
                    actionRow("arrow.clockwise", "actions.requestData", "actions.requestDataSub") {
                        requestData()
                    }
                    actionRow("doc.on.doc", "actions.copyAgentId", nil) {
                        UIPasteboard.general.string = agent.id
                        showResult(tr("actions.agentIdCopied"))
                    }
                    if agent.permissionLevel >= 3 {
                        actionRow("power", "actions.reboot", "actions.rebootSub", danger: true, divider: false) {
                            confirmReboot = true
                        }
                    }
                }
                .padding(16)

                if let busyMessage {
                    HStack(spacing: 10) { ProgressView(); Text(busyMessage).foregroundColor(t.fg2) }
                        .font(.system(size: 13)).padding(.horizontal, 20)
                } else if let resultMessage {
                    HStack(spacing: 9) {
                        Image(systemName: resultIsError ? "xmark.circle.fill" : "checkmark.circle.fill")
                            .foregroundColor(resultIsError ? t.crit : t.ok)
                        Text(resultMessage).font(.system(size: 13)).foregroundColor(t.fg2)
                    }
                    .padding(.horizontal, 20)
                }
                Spacer()
            }
            .background(t.bg)
            .navigationTitle(tr("actions.title"))
            .navigationBarTitleDisplayMode(.inline)
            .toolbar { ToolbarItem(placement: .confirmationAction) { Button(tr("common.dismiss")) { dismiss() } } }
        }
        .presentationDetents([.medium])
        .presentationDragIndicator(.visible)
        .alert(tr("actions.rebootConfirmTitle"), isPresented: $confirmReboot) {
            Button(tr("common.cancel"), role: .cancel) { }
            Button(tr("actions.rebootConfirm"), role: .destructive) { reboot() }
        } message: { Text(tr("actions.rebootConfirmBody")) }
    }

    private func actionRow(_ icon: String, _ title: String, _ sub: String?,
                           danger: Bool = false, divider: Bool = true,
                           action: @escaping () -> Void) -> some View {
        NanoListRow(divider: divider, onTap: busyMessage == nil ? action : nil) {
            VStack(alignment: .leading, spacing: 2) {
                Text(tr(title)).font(.system(size: 15, weight: .medium)).foregroundColor(danger ? t.crit : t.fg)
                if let sub { Text(tr(sub)).font(.system(size: 12.5)).foregroundColor(t.fg3) }
            }
        } leading: {
            Image(systemName: icon).font(.system(size: 18)).foregroundColor(danger ? t.crit : t.accent).frame(width: 24)
        } trailing: {
            Image(systemName: "chevron.right").font(.system(size: 12)).foregroundColor(t.fg4)
        }
    }

    private func requestData() {
        guard let service = store.serviceForAgent(agent.id) else {
            showResult(tr("actions.noServerForNode"), error: true); return
        }
        busyMessage = tr("actions.requestDataPending")
        resultMessage = nil
        Task {
            let error = await service.requestData(agent.id)
            busyMessage = nil
            if let error { showResult(tr("actions.dataRequestFailed", ["error": error]), error: true) }
            else { showResult(tr("actions.dataRequested")) }
        }
    }

    private func reboot() {
        guard let service = store.serviceForAgent(agent.id) else {
            showResult(tr("actions.noServerForNode"), error: true); return
        }
        busyMessage = tr("actions.rebootPending")
        resultMessage = nil
        Task {
            let dispatch = await service.sendCommandReturningId(agent.id, type: "SYSTEM_REBOOT")
            guard dispatch.ok, let commandId = dispatch.commandId else {
                busyMessage = nil
                showResult(tr("actions.rebootFailed", ["error": dispatch.error ?? "unknown"]), error: true)
                return
            }
            for _ in 0..<6 {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                let result = await service.pollCommandResult(agent.id, commandId)
                switch result.status {
                case .ready:
                    busyMessage = nil; showResult(tr("actions.rebootDone")); return
                case .denied:
                    busyMessage = nil; showResult(tr("actions.commandDenied"), error: true); return
                case .error:
                    busyMessage = nil; showResult(tr("actions.rebootFailed", ["error": result.message ?? "unknown"]), error: true); return
                case .pending: continue
                }
            }
            busyMessage = nil
            showResult(tr("actions.rebootSent"))
        }
    }

    private func showResult(_ message: String, error: Bool = false) {
        resultMessage = message
        resultIsError = error
        Task {
            try? await Task.sleep(nanoseconds: 6_000_000_000)
            if resultMessage == message { resultMessage = nil }
        }
    }
}
